docs/rfcs/2024-01-17-dataflow-framework.md

```diff
@@ -0,0 +1,57 @@
+---
+Feature Name: Dataflow Framework
+Tracking Issue: 
+Date: 2024-01-17
+Author: "Discord9 <discord9@163.com>"
+---
+
+# Summary
+This RFC proposes a Lightweight Module for executing continuous aggregation queries on a stream of data.
+
+# Motivation
+Being able to do continuous aggregation is a very powerful tool. It allows you to do things like:
+1. downsample data from i.e. 1 milliseconds to 1 second
+2. calculate the average of a stream of data
+3. Keeping a sliding window of data in memory
+In order to do those things while maintaining a low memory footprint, you need to be able to manage the data in a smart way. Hence, we only store necessary data in memory, and send/recv data deltas to/from the client.
+# Details
+## System boundary / What it's and isn't
+- GreptimeFlow provides a way to perform continuous aggregation over time-series data.
+- It's not a complete streaming-processing system. Only a must subset functionalities are provided.
+- Flow can process a configured range of fresh data. Data exceeding this range will be dropped directly. Thus it cannot handle random datasets (random on timestamp).
+- Both sliding windows (e.g., latest 5m from present) and fixed windows (every 5m from some time) are supported. And these two are the major targeting scenarios
+- Flow can handle most aggregate operators within one table(i.e. Sum, avg, min, max and comparison operators). But others (join, trigger, txn etc.) are not the target feature.
+Framework
+- Greptime Flow's is built on top of [Hydroflow](https://github.com/hydro-project/hydroflow).
+- We have three choices for the Dataflow/Streaming process framework for our simple continuous aggregation feature:
+1. Based on the timely/differential dataflow crate that [materialize](https://github.com/MaterializeInc/materialize) based on. Later, it's proved too obscure for a simple usage, and is hard to customize memory usage control.
+2. Based on a simple dataflow framework that we write from ground up, like what [arroyo](https://www.arroyo.dev/) or [risingwave](https://www.risingwave.dev/) did, for example the core streaming logic of [arroyo](https://github.com/ArroyoSystems/arroyo/blob/master/arroyo-datastream/src/lib.rs) only takes up to 2000 line of codes. However, it means maintaining another layer of dataflow framework, which might seem easy in the beginning, but I fear it might be too burdensome to maintain once we need more features.
+3. Based on a simple and lower level dataflow framework that someone else write, like [hydroflow](https://github.com/hydro-project/hydroflow), this approach combines the best of both worlds. Firstly, it boasts ease of comprehension and customization. Secondly, the dataflow framework offers precisely the necessary features for crafting uncomplicated single-node dataflow programs while delivering decent performance.
```

Comment:
Looks like hydroflow is still a young research project. I think we should add a middle layer to abstract the engine just like we are using DataFusion. It would be possible If we want to change the engine in future.